Fix inventory plugin broken with AzureStackHCI vms and also fix ScaleSets network interfaces

TiTi opened the pull request:
##### SUMMARY
Fix of #2074

- Prevents failure if inventory also contains AzureStackHCI vms (or just arc vms)
- Remove duplicate calls to instance view REST endpoints
- Fix original problem when retreiving scale sets vms network interface properties
=> will provide scale set private ip under private_ipv4_addresses host var
- Simplify logic

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
inventory plugin

##### ADDITIONAL INFORMATION

The original issue was : `_on_nic_response` was never called for scale sets nics, despite call to `inventory_client._enqueue_get(url=nic['id'],` being made

To reproduce issue:
`az rest --url "/subscriptions/<subscription_id>/resourceGroups/<rgName>/providers/Microsoft.Compute/virtualMachineScaleSets/<scaleSetName>/virtualMachines/<vmId>/networkInterfaces/<NicName>?api-version=2024-05-01"`

Which returns:
Bad Request({"error":{"code":"NoRegisteredProviderFound","message":"No registered resource provider found for location 'westeurope' and API version '2024-05-01' for type 'virtualMachineScaleSets/virtualMachines/networkInterfaces'. The supported api-versions are '2015-05-01-preview, 2015-06-15, 2016-03-30, 2016-04-30-preview, 2016-06-01, 2016-07-01, 2016-08-01, 2016-09-01, 2017-03-30, 2017-12-01, 2018-04-01, 2018-06-01, 2018-10-01, 2019-03-01, 2019-07-01, 2019-12-01, 2020-06-01, 2020-12-01, 2021-03-01, 2021-04-01, 2021-07-01, 2021-11-01, 2022-03-01, 2022-08-01, 2022-11-01, 2023-03-01, 2023-07-01, 2023-09-01, 2024-03-01, 2024-07-01, 2024-11-01, 2025-04-01'. The supported locations are 'eastus, eastus2, westus, centralus, northcentralus, southcentralus, northeurope, westeurope, eastasia, southeastasia, japaneast, japanwest, australiaeast, australiasoutheast, australiacentral, brazilsouth, southindia, centralindia, westindia, canadacentral, canadaeast, westus2, westcentralus, uksouth, ukwest, koreacentral, koreasouth, francecentral, southafricanorth, uaenorth, switzerlandnorth, germanywestcentral, norwayeast, jioindiawest, westus3, swedencentral, qatarcentral, polandcentral, italynorth, israelcentral, spaincentral, mexicocentral, malaysiawest, newzealandnorth, indonesiacentral, austriaeast, chilecentral'."}})

TiTi committed:
Quick fix of issue #2074

TiTi committed:
Better fix

TiTi committed:
Fix type

TiTi committed:
Remove duplicate call to instanceView

Everything is already available in self._instanceview

Note: changes os_compute_name to computer_name

TiTi committed:
Fix test

TiTi committed:
Fix the original problem : incorrect api version for scaleSet

to reproduce issue:
az rest --url "/subscriptions/<subscription_id>/resourceGroups/<rgName>/providers/Microsoft.Compute/virtualMachineScaleSets/<scaleSetName>/virtualMachines/777/networkInterfaces/priva5b52Nic?api-version=2025-04-01"

returns:
Bad Request({"error":{"code":"NoRegisteredProviderFound","message":"No registered resource provider found for location 'westeurope' and API version '2025-05-01' for type 'virtualMachineScaleSets/virtualMachines/networkInterfaces'. The supported api-versions are '2015-05-01-preview, 2015-06-15, 2016-03-30, 2016-04-30-preview, 2016-06-01, 2016-07-01, 2016-08-01, 2016-09-01, 2017-03-30, 2017-12-01, 2018-04-01, 2018-06-01, 2018-10-01, 2019-03-01, 2019-07-01, 2019-12-01, 2020-06-01, 2020-12-01, 2021-03-01, 2021-04-01, 2021-07-01, 2021-11-01, 2022-03-01, 2022-08-01, 2022-11-01, 2023-03-01, 2023-07-01, 2023-09-01, 2024-03-01, 2024-07-01, 2024-11-01, 2025-04-01'. The supported locations are 'eastus, eastus2, westus, centralus, northcentralus, southcentralus, northeurope, westeurope, eastasia, southeastasia, japaneast, japanwest, australiaeast, australiasoutheast, australiacentral, brazilsouth, southindia, centralindia, westindia, canadacentral, canadaeast, westus2, westcentralus, uksouth, ukwest, koreacentral, koreasouth, francecentral, southafricanorth, uaenorth, switzerlandnorth, germanywestcentral, norwayeast, jioindiawest, westus3, swedencentral, qatarcentral, polandcentral, italynorth, israelcentral, spaincentral, mexicocentral, malaysiawest, newzealandnorth, indonesiacentral, austriaeast, chilecentral'."}})

TiTi committed:
Revert to same logic to parse scale set networkinterface

Basically a revert of PR #2055 now that proper api_version is specific

Fred-sun commented on plugins/inventory/azure_rm.py:
```diff
@@ -823,7 +833,7 @@ def hostvars(self):
             location=self._arcvm['location'] if self._arcvm else self._vm_model['location'],
             name=self._arcvm['name'] if self._arcvm else self._vm_model['name'],
-            computer_name=self._vm_model['properties'].get('osProfile', {}).get('computerName'),
+            computer_name=self._instanceview['computerName'] if self._instanceview.get('computerName') else self._vm_model['properties'].get('osProfile', {}).get('computerName'),
```
line too long (178 > 160 characters)

TiTi replied:
fixed :)

plus i've added some logging :) :) :) check out with `-vvvv` or `debug=true` in your ansible.cfg

TiTi committed:
Display when errors in responses

Also display urls called with -vvvv
And all raw responses if debug=true in ansible.cfg

TiTi committed:
Fix line too long

TiTi committed:
Display request causing error to help understand what's going on

TiTi commented:
any update @Fred-sun ?

TiTi committed:
Merge branch 'dev' into fix-inventory-plugin

Fred-sun commented on plugins/inventory/azure_rm.py:
```diff
@@ -187,6 +186,8 @@
 
 
+display = Display()
+
```
expected 2 blank lines, found 1

TiTi replied:
it's fixed

TiTi committed:
Expects 2 blank lines

zunyangc commented:
Hi @TiTi , sorry for the slow progress on this PR. As I currently still do not have access to a pre-deployed Azure Stack HCI VM that is registered via Azure Arc, I couldn't validate this PR, hope for your understanding. 

Let me discuss with the team and see how we can proceed further. Thanks!

TiTi committed:
Merge remote-tracking branch 'origin/dev' into fix-inventory-plugin

# Conflicts:
#	plugins/inventory/azure_rm.py

Klaas- commented:
so general answer, os_computer_name / computer_name I personally don't care too much about them because I am not using them from inventory :D from a compatibility perspective I would say don't break existing behavior unless the MSFT people here are up for a major version increase.

Now if you ask me what should happen with the inventory generally speaking, I think what we do here should not be done here :) We should not be making up our own names for variables and completely rely on the azure python sdk, I've opened a "big issue" for that https://github.com/ansible-collections/azure/issues/2130 . Having said all that, this is not going to be short term solution because I guess I have to first convince msft of actually documenting the batch endpoint, then getting it into the sdk, then we can start using it :)

For shortterm changes, the smaller you can make your PRs the easier it is to merge them for the MSFT people, stuff like HCI and/or ARC is a problem because they don't have this in their test environment apparently...

zunyangc commented:
Hi @TiTi & @Klaas- , thanks again for the work on this PR.

During verification, I found an additional issue related to Azure Stack HCI machines.

When an HCI VM is onboarded, the `properties.status` field is not always a string (e.g. when awaiting connection). Depending on the API version and onboarding state, it may return a structure like:
````
 {'id': '/subscriptions/<hidden>/resourceGroups/ansible-test-2077/providers/Microsoft.HybridCompute/machines/HCIVMansible2077', 'name': 'HCIVMansible2077', 'location': 'australiaeast', 'type': 'Microsoft.HybridCompute/machines', 'tags': {}, 'identity': {'type': 'SystemAssigned', 'principalId': '<hidden>', 'tenantId': '<hidden>'}, 'properties': {'provisioningState': 'Succeeded', 'osProfile': {}, 'vmId': '<hidden>', 'agentConfiguration': {}, 'agentUpgrade': {'enableAutomaticUpgrade': False}, 'licenseProfile': {'esuProfile': {'licenseAssignmentState': 'NotAssigned', 'serverType': 'Unknown', 'esuEligibility': 'Unknown', 'esuKeyState': 'Inactive'}}}, 'kind': 'HCI'}
````
Because the current implementation does:
````
self._status = self._arc_model['properties'].get('status', {}).lower()
````
that will result in a runtime failure:
````
'dict' object has no attribute 'lower'
````

Other than that, LGTM. 

Thanks!

Klaas- commented:
So I am guessing you want to change that to:
`self._status = self._arc_model['properties'].get('status', '').lower()`

so the lower always works?
